### DISCUSSION TITLE
Indexer Processing over Large Transaction History

### DISCUSSION THREAD

**Discussion by l-monninger** [movementlabsxyz/movement#432] [Ideas]
# Summary
We may have discovered that an exposed full node cannot handle live operations and provide the stream needed to synchronize over a lengthy history. For example, running the indexer from the start on the testnet which has 50 million transactions has proved problematic.

The medium-term solution here is horizontal scaling. However, we have also discussed stream metering.

**Comment by l-monninger** [comment]
Per concerns about establishing new connections:

>  we do have nginx as reverse proxy for the suzuka node ... so it should optimise the connection

The HTTP/2 connection should be kept alive regardless of Nginx, I believe. But, I believe gRPC processing a new stream connection is a bit heavier...

**Reply by l-monninger** [reply]
Eh, I mean, its nothing too heavy. You do have Tokio runtime context to setup and basic gRPC stream logical operations. But, fairly reasonable.

**Reply by musitdev** [reply]
One other point is when the indexer has synchronized some Tx, it restarts after the last synchronization point. So the initial dump can be done little by little.

**Reply by musitdev** [reply]
This is the code use to connect:
```
        Ok(client) => client
            .accept_compressed(tonic::codec::CompressionEncoding::Gzip)
            .accept_compressed(tonic::codec::CompressionEncoding::Zstd)
            .send_compressed(tonic::codec::CompressionEncoding::Zstd)
            .max_decoding_message_size(MAX_RESPONSE_SIZE)
            .max_encoding_message_size(MAX_RESPONSE_SIZE),
```
and to fetch a batch ot Tx:
```
    let mut request = tonic::Request::new(GetTransactionsRequest {
        starting_version: Some(starting_version),
        transactions_count,
        ..GetTransactionsRequest::default()
    });
    request.metadata_mut().insert(
        GRPC_API_GATEWAY_API_KEY_HEADER,
        format!("Bearer {}", grpc_auth_token.clone())
            .parse()
            .unwrap(),
    );
    request
        .metadata_mut()
        .insert(GRPC_REQUEST_NAME_HEADER, processor_name.parse().unwrap());

client.get_transactions(request).await
```

**Reply by l-monninger** [reply]
So, do you think making the `transactions_count` a little larger and maybe adding an option to sleep in between a little could help?

**Reply by musitdev** [reply]
I'm not sure that increasing the Tx count will change the load on the suzuka node. I think it's already a parameter of the conf. You extract more Tx in one request (so more load on the node) but you take more time to send it (so less load). In the end, we can still exceed load level that break the node.
On the other hand, adding some timing during 2 queries to avoid to always put on load the node can help to avoid crashing the node. It will increase the synchronization time, but It's not an issue.
It's not complicated. I can add the parameter: SLEEP_TIME_BETWENN_REQUEST_MS that define the sleep time in ms between 2 request. I set it by default to 10 ms.
I was thinking, the size of the range is in version. Version are roughly the height, so 4 versions of range mean 4 blocks.
I try to do that now and update the indexer rev version on movement.

**Reply by musitdev** [reply]
I've created 2 PR that add a sleep time between 2 gRpc request:
The Aptos indexer project: https://github.com/movementlabsxyz/aptos-indexer-processors/pull/3
The Movement project that update the indexer server: https://github.com/movementlabsxyz/movement/pull/436

If you only update the indexer with PR #436 it should work without updating the Suzuka node.
To change the sleep time from 10ms to another value, set the env var: SLEEP_TIME_BETWENN_REQUEST_MS with the duration in ms.
